File: src/main/java/org/taktik/icure/security/AuthenticationFailureHandler.java
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.security;

import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class AuthenticationFailureHandler extends SimpleUrlAuthenticationFailureHandler {
	public static final String FAILURE_URL = "failureURL";

	@Override
	public void onAuthenticationFailure(HttpServletRequest httpRequest, HttpServletResponse httpResponse, AuthenticationException exception) throws IOException, ServletException {
		String failureURL = (String) httpRequest.getSession().getAttribute(FAILURE_URL);

		if (failureURL == null) {
			super.onAuthenticationFailure(httpRequest, httpResponse, exception);
		} else {
			saveException(httpRequest, exception);
			logger.debug("Redirecting to " + failureURL);
			httpResponse.sendRedirect(failureURL);
		}
	}
}

File: src/main/kotlin/org/taktik/icure/security/AuthenticationFailureHandler.kt
/*
 * Copyright (c) 2020. Taktik SA, All rights reserved.
 */
package org.taktik.icure.security

import java.io.IOException
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler

class AuthenticationFailureHandler : SimpleUrlAuthenticationFailureHandler() {
	@Throws(IOException::class, ServletException::class)
	override fun onAuthenticationFailure(
		httpRequest: HttpServletRequest,
		httpResponse: HttpServletResponse,
		exception: AuthenticationException
	) {
		val failureURL = httpRequest.session.getAttribute(FAILURE_URL) as String?
		if (failureURL == null) {
			super.onAuthenticationFailure(httpRequest, httpResponse, exception)
		} else {
			saveException(httpRequest, exception)
			logger.debug("Redirecting to $failureURL")
			httpResponse.sendRedirect(failureURL)
		}
	}

	companion object {
		const val FAILURE_URL = "failureURL"
	}
}
